File: cdm-sources/src/cdm/base/datetime/functions/BusinessCenterHolidaysMultiple.py
# pylint: disable=line-too-long, invalid-name, missing-function-docstring, missing-module-docstring, superfluous-parens
# pylint: disable=wrong-import-position, unused-import, unused-wildcard-import, wildcard-import, wrong-import-order, missing-class-docstring
from __future__ import annotations
import sys
import datetime
import inspect
from decimal import Decimal
from rosetta.runtime.utils import *
from rosetta.runtime.func_proxy import replaceable, create_module_attr_guardian
from cdm.base.datetime.BusinessCenterEnum import BusinessCenterEnum
from cdm.base.datetime.functions.BusinessCenterHolidays import BusinessCenterHolidays

__all__ = ['BusinessCenterHolidaysMultiple']


@replaceable
def BusinessCenterHolidaysMultiple(businessCenters: list[BusinessCenterEnum] | None) -> datetime.date:
    """
    Returns a merged list of holidays for the supplied business centers.
    
    Parameters 
    ----------
    businessCenters : BusinessCenterEnum
    The business centers for which the holiday list is required.
    
    Returns
    -------
    holidayDates : date
    
    """
    self = inspect.currentframe()
    
    
    holidayDates = (lambda item: sorted(item))((lambda item: set(item))((lambda item: flatten_list(item))(map(lambda item: BusinessCenterHolidays(item), rosetta_resolve_attr(self, "businessCenters")))))
    
    
    return holidayDates

sys.modules[__name__].__class__ = create_module_attr_guardian(sys.modules[__name__].__class__)

File: cdm-sources/cdm/base/datetime/functions/BusinessCenterHolidaysMultiple.java
package cdm.base.datetime.functions;

import cdm.base.datetime.BusinessCenterEnum;
import com.google.inject.ImplementedBy;
import com.rosetta.model.lib.functions.RosettaFunction;
import com.rosetta.model.lib.mapper.MapperC;
import com.rosetta.model.lib.mapper.MapperListOfLists;
import com.rosetta.model.lib.records.Date;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import javax.inject.Inject;

import static com.rosetta.model.lib.expression.ExpressionOperators.*;

@ImplementedBy(BusinessCenterHolidaysMultiple.BusinessCenterHolidaysMultipleDefault.class)
public abstract class BusinessCenterHolidaysMultiple implements RosettaFunction {
	
	// RosettaFunction dependencies
	//
	@Inject protected BusinessCenterHolidays businessCenterHolidays;

	/**
	* @param businessCenters The business centers for which the holiday list is required.
	* @return holidayDates The combined list of holidays in all of the supplied business centers.
	*/
	public List<Date> evaluate(List<BusinessCenterEnum> businessCenters) {
		List<Date> holidayDates = doEvaluate(businessCenters);
		
		return holidayDates;
	}

	protected abstract List<Date> doEvaluate(List<BusinessCenterEnum> businessCenters);

	public static class BusinessCenterHolidaysMultipleDefault extends BusinessCenterHolidaysMultiple {
		@Override
		protected List<Date> doEvaluate(List<BusinessCenterEnum> businessCenters) {
			if (businessCenters == null) {
				businessCenters = Collections.emptyList();
			}
			List<Date> holidayDates = new ArrayList<>();
			return assignOutput(holidayDates, businessCenters);
		}
		
		protected List<Date> assignOutput(List<Date> holidayDates, List<BusinessCenterEnum> businessCenters) {
			final MapperListOfLists<Date> thenArg0 = MapperC.<BusinessCenterEnum>of(businessCenters)
				.mapItemToList(item -> MapperC.<Date>of(businessCenterHolidays.evaluate(item.get())));
			final MapperC<Date> thenArg1 = thenArg0
				.flattenList();
			final MapperC<Date> thenArg2 = distinct(thenArg1);
			holidayDates.addAll(thenArg2
				.sort().getMulti());
			
			return holidayDates;
		}
	}
}
